src/flask_pyoidc/flask_pyoidc.py

```diff
@@ -319,3 +334,205 @@ def valid_access_token(self, force_refresh=False):
                        refresh_token=response.get('refresh_token'))
         return access_token
+
+    def _check_authorization_header(self, request) -> bool:
+        '''Look for authorization in request header.
+
+        Parameters
+        ----------
+        request : werkzeug.local.LocalProxy
+            flask request object.
+
+        Returns
+        -------
+        bool
+            True if the request header contains authorization else False.
+        '''
+        if 'Authorization' in request.headers and request.headers['Authorization'].startswith('Bearer '):
+            return True
+        logger.info('missing authorization field')
+        return False
+
+    def _parse_access_token(self, request) -> str:
+        '''Parse access token from the authorization request header.
+
+        Parameters
+        ----------
+        request : werkzeug.local.LocalProxy
+            flask request object.
+
+        Returns
+        -------
+        accept_token : str
+            access token from the request header.
+        '''
+        _, access_token = request.headers['Authorization'].split(maxsplit=1)
+        return access_token
+
+    def introspect_token(self, request, client, scopes: list = None) -> TokenIntrospectionResponse:
+        '''RFC 7662: Token Introspection
+        The Token Introspection extension defines a mechanism for resource
+        servers to obtain information about access tokens. With this spec,
+        resource servers can check the validity of access tokens, and find out
+        other information such as which user and which scopes are associated
+        with the token.
+
+        Parameters
+        ----------
+        request : werkzeug.local.LocalProxy
+            flask request object.
+        client : flask_pyoidc.pyoidc_facade.PyoidcFacade
+            PyoidcFacdae object contains metadata of the provider and client.
+        scopes : list
+            Specify scopes that are bound for the end endpoint.
+
+        Returns
+        -------
+        bool
+            True if access_token is valid else False.
+
+        Raises
+        ------
+        NotForMe
+            If access_token is invalid.
+
+        '''
+        received_access_token = self._parse_access_token(request)
+        # send token introspection request
+        result = client._token_introspection_request(
+            access_token=received_access_token)
+        logger.debug(result)
+        # Check if access_token is valid, active can be True or False
+        if not result.get('active'):
+            return False
+        # Check if client_id is in audience claim
+        if not client._client.client_id in result['aud']:
+            logger.info('Token is valid but required audience is missing')
+            # raises exception if client_id is not in audience, you can
+            # configure audience from Identity Provider
+            raise NotForMe('required audience is missing')
```

Comment:
This should raise an exception that is propagated to user code.
I think it would suffice to log it (probably info level) and return `False` instead.

Reply:
If we don't raise exception, it will be difficult for the user to debug why authorization is failing despite the presence of a valid token. flask.abort 403 & 401 do not give necessary information about the failure. These failures will go undetected in monitoring dashboard that health checks microservices. Most users don't enable logging for packages so what went wrong won't show up on the console either.

Reply:
My main reason for not raising an exception is that then will this require users of this extension to handle it themselves, instead of having it work right out of the box.
And in general it's a good thing to expose as little information as possible from the API itself to avoid it somehow being misused by attackers.

But a custom message might be helpful, however I think it should be even more generic.
Maybe something like: 'You do not have sufficient privileges for access.'

Reply:
Let's keep the default message then. It's good enough. I'll document possible failures in the documentation.